Support conditional and optional includes

scheibelp opened the pull request:
UPDATE: Instead of the original idea of supporting `include:` as a subsection of `config:`, this PR has changed to support `include:` as an independent schema while retaining the new capabilities of supporting conditional, unconditional, and optional includes.  The features are supported through `spack:include:` with essentially the same include syntax, i.e.:

```
include:
- path: "/path/to/a/config-dir-or-file"
  when: os == "ventura"
- "/another/path/added/unconditionally"
- path: "/base/dir/with/spack/variables/$os/$target/"
  optional: true
```

This PR also supports conditional and optional includes for environment include paths.

TODO

- [x] replace `config: include:` with `include:` support
- [x] Update docs
- [x] support URLs for global includes (as has been supported by environment includes)
- [x] support relative paths for global includes (relative to the source path)
- [x] resolve issues with unit tests
- [x] Change `spack commands` (fish scopes) completion message to note that user's includes need to be left off
- [x] resolve circular import issue introduced with above change(s)
- [x] Revise docs as needed
- [x] Move the commits to a separate PR as requested (#48784)

--------
@greenc-FNAL @tgamblin 

This adds an `include:` subsection to `config:`.  The `include:` subsection specifies a list of config paths (either directories or single files) that should be added as scopes to the configuration. For example:

```
config:
  include:
  - path: "/path/to/a/config-dir-or-file"
    when: os == "ventura"
  - "/another/path/added/unconditionally"
  - path: "/base/dir/with/spack/variables/$os/$target/"
    optional: true
```

If this were part of the site config scope, or added with `spack -C`, then all Spack commands would run with the config from `/path/to/a/config-dir-or-file` (e.g. `spack config blame`) when on Mac Ventura.

Inclusion is currently not recursive: if you include another `config.yaml` that itself defines includes, you will not get those includes.

This includes support for conditional activation of configs, using `when`, which functions like `when:` for spec list definitions (see: https://spack.readthedocs.io/en/latest/environments.html#spec-list-references). The `when` string is `eval`ed, so can support something like `os == "ventura" and target == ...`.

Altogether, this is intended as a more general alternative to https://github.com/spack/spack/pull/40018: you could conditionally activate a config for each os and/or target.

scheibelp committed:
Initial pass to support config:include:
 - partial include logic as subsection of config: (scheibelp 01f69fb17e60da973c3d9402817b554123ddf891)
 - ref error (scheibelp 161097ab3198049f56312de8fbdcae8e461b01ca)
 - refactor function used only for env encludes to support importing other lists of paths as config scopes (scheibelp e3b0da247fd48989b584bcd9bf36f063ea14f6eb)
 - initial test for included configs (scheibelp ae1aaa9065ff61979edf3d04cc468967c1b8512c)
 - style fix (scheibelp a54e5651339d3d4f5ae39933dc079a8a5d7957ba)
 - expanded test with additional entries (scheibelp a6795af237492feb609d50fcc2962f3f428494ed)
 - add docstring (scheibelp a18952e2b19e48dbe6da0895bc8ca037ad60f968)
 - all commands now make use of config:includes (scheibelp bd666d82345fbd5ecfc6d19fd3938a1104eeb77f)
 - includes -> include; file -> path; also indicate in schema that additional properties are not allowed (scheibelp b9af793acf09de3bd0cb9c5b44940dfa4ef0f417)
 - allow include: entries to be strings if they don't have an associated 'when' (scheibelp 5899d453dd5a23fd8d0b2e5e5e7e629317d7cd83)
 - auto style fix (scheibelp c9bb265c61e6a195e4cb5f6856a64958da4e96e6)
 - avoid repetition in test logic (scheibelp fe27fb3665fcd5789987b24a9bc4dea153d24e30)
 - now that include: has a path-only option, 'when' is always required when specifying 'path' (scheibelp f635b8e906623fee56bc345082e83876132337f7)
 - enforce requiring 'path' and 'when' at the schema level (scheibelp 5e07209819f0e7b36cef4d2fd57e20ca9560bb2c)
 - now that all used properties must be mentioned, fix an unrelated test (scheibelp 47684509b38d67439138522d8d1a84afb06172e5)
 - rewrite test in terms of a config section that actually exists (scheibelp 3fae4d422df34d8cd112e61aa5fe931cc354c8be)
 - other test depending on (scheibelp made-up) config:info refactored (da281022f9f8764fc24821104271a0e1b5409c3d)
 - test was somehow passing before based on now-deprecated reference to install_tree (previously install_root) - failing now that I disallow made-up properties in config (scheibelp 8a04e0eaa0164ea5949c2dfd72cda21d7c32f591)
 - partial testing of canonicalized include paths and optional includes (scheibelp e376e9191bd6cb7f6c6f5437a58483963bb04340)
 - better testing of prior change; added explanatory comments to test (scheibelp X60671a2d8e1a18c90f91a511235bb9a08496059c)
 - auto style fix (scheibelp 6787336b68dfdaff3e33f6182d9d847c90bed651)
 - update_config_with_includes: fix section name to match change (tldahlgren 444819ba9c48c0836c1d2b48c41c177f8cd67668)
 - scopes_from_paths: added docstring and type hints; converted to using f-strings (tldahlgren ff5fe7b96b9a34f78751b656af127066981f8548)

tgamblin commented on lib/spack/spack/schema/config.py:
```diff
@@ -94,6 +94,14 @@
             "additional_external_search_paths": {"type": "array", "items": {"type": "string"}},
             "binary_index_ttl": {"type": "integer", "minimum": 0},
             "aliases": {"type": "object", "patternProperties": {r"\w[\w-]*": {"type": "string"}}},
+            "includes": {
```
Can we please call this `include:` to match the existing env `include:`?https://spack.readthedocs.io/en/latest/environments.html#included-configurations

tgamblin commented on lib/spack/spack/schema/config.py:
```diff
@@ -94,6 +94,14 @@
+            "includes": {
+                "type": "array",
+                "default": [],
+                "items": {
```
Should also allow a string to be an element of items, to match existing env include. e.g.:

```yaml
include:
- path/to/first/config
- path/to/second/config
```

Would be the same as:
```yaml
include:
- path: path/to/first/config
- path: path/to/second/config
```

tgamblin commented on lib/spack/spack/schema/config.py:
```diff
@@ -94,6 +94,14 @@
+                "default": [],
+                "items": {
+                    "type": "object",
+                    "properties": {"when": {"type": "string"}, "file": {"type": "string"}},
```
This says `file:` but the PR description says `path:` -- I like `path:` better as we could include either a single config file or a path to a config directory.

tgamblin reviewed (changes requested):
This looks awesome to me.

I think the PR description is not quite specific enough:

> If this were part of the site config scope, or added with spack -C, then all Spack commands would run with the config from /path/to/a/config-dir-or-file (e.g. spack config blame).

I think it needs "when on macOS Ventura".

Other than that a few cosmetic requests below.

greenc-FNAL commented:
Thanks for this, @scheibelp. I've asked the rest of the team for their comments, but my $0.02:

At first glance, this looks great, thank you! My one thought is that a Spack config for a CVMFS area supporting multiple OSes and/or targets could get pretty verbose unless we were able to have evaluable paths (e.g. `/path/to/config/$os`). That feature would also probably need to come with the ability to specify optional includes that inform rather than raise an exception if the file or directory is not present as calculated for the executing system.

scheibelp commented:
> At first glance, this looks great, thank you! My one thought is that a Spack config for a CVMFS area supporting multiple OSes and/or targets could get pretty verbose unless we were able to have evaluable paths (e.g. /path/to/config/$os). 

That sounds like a good idea: for this PR in its current form you would need a long list of `include:` to address your use case, but something like:

```
- include:
  - path: /a/b/c/$os/$target
    skip_if_nonexistent: true
```

would allow for a compact include file. That will take me a couple days to address here.

scheibelp commented:
>> My one thought is that a Spack config for a CVMFS area supporting multiple OSes and/or targets could get pretty verbose unless we were able to have evaluable paths (e.g. /path/to/config/$os). 

> That sounds like a good idea: for this PR in its current form you would need a long list of `include:` to address your use case, but something like:

That is now supported (and tested) - that took less time than I originally assumed.

tgamblin commented:
Requesting a review from @tldahlgren as I think she should give this a glance to see if it introduces any tricky config layering issues. She's familiar with these from recently working on #33768.

tldahlgren commented on lib/spack/spack/config.py:
```diff
@@ -764,6 +764,51 @@ def _add_platform_scope(
+def update_config_with_includes():
+    """The "config:" section of a Configuration can specify other
+    configurations to include. This does not handle recursive includes
+    (i.e. if an included config defines an "includes:" section).
```
```suggestion
    (i.e. if an included config defines an "include:" section).
```

tldahlgren commented on lib/spack/spack/environment/environment.py:
```diff
@@ -3113,6 +3033,90 @@ def use_config(self):
+                    tty.warn(
+                        "Will not re-stage configuration from {0} to avoid "
+                        "losing changes to the already staged file of the "
+                        "same name.".format(remote_path)
```
Since the code is being moved, should be using f-strings (here and elsewhere in the method).

greenc-FNAL commented:
@scheibelp Does this only allow `include` for `config` items, or also for e.g. `packages`,_ et al_?

tgamblin commented:
@greenc-FNAL the plan is to make `include:` its own top-level thing, not `config:include:`

scheibelp commented:
> Does this only allow include for config items, or also for e.g. packages, compilers, et al?

> @greenc-FNAL the plan is to make include: its own top-level thing, not config:include:

It its current form, you can include any config section you want: an `include:`d path can for example refer to a directory with a `packages.yaml`, `repos.yaml`, etc. It can also refer to a merged configuration that is stored in a single file (which can have a `packages:` section, a `repos:` section, and so on).

That will also be true when it is updated as Todd mentions.

tldahlgren committed:
Schema: separate 'include' into own section
- initial separation (tldahlgren 34d96c69b3dbdf192db621d3bf8018d973248b36)
- include schema: fix starting line in file's docstring (tldahlgren 83346673d6d079d3ab68ef030f3648c16c1123f5)
- Environment schema: (re)use new include schema (tldahlgren a303d5e735ce592c248bf6e8b97bcfd45e6bd9b4)

tldahlgren committed:
remove support for config:include
- initial removal (tldahlgren 296971c041a46223fba7ae7f29bbcf1ba57a1f62)
- Restore unrelated test bug fixed in ccb69cc (tldahlgren 887ec635cde2198292bf94548018dca99a695b11)

tldahlgren committed:
Environments: support the new include schema

tldahlgren committed:
Post-rebase cleanup

tldahlgren committed:
Docs/environment includes: update to include new optional and conditional examples

tldahlgren reviewed (dismissed):
Clearing my `Request changes` review.

tldahlgren committed:
Fix style issue with env schema

tgamblin commented on lib/spack/spack/schema/config.py:
```diff
@@ -19,6 +19,7 @@
     "config": {
         "type": "object",
         "default": {},
+        "additionalProperties": False,
```
I think we shouldn't add this here, as we do remove config options that we don't want to fail on.  Probably a bigger discussion whether we should use or not use this.

tgamblin commented on lib/spack/spack/environment/environment.py:
```diff
@@ -2933,95 +2933,43 @@ def included_config_scopes(self) -> List[spack.config.ConfigScope]:
-
-        # load config scopes added via 'include:', in reverse so that
-        # highest-precedence scopes are last.
         includes = self[TOP_LEVEL_KEY].get("include", [])
```
I think this should use `spack.config.get` now, right?  Otherwise it may not get the full list.

Actually, it looks like this whole method should just be deleted out of environments and the handling for it should be done in `spack.config` -- as the PR is right now, included config is *only* processed in environments.

tldahlgren replied:
>  I think this should use spack.config.get now, right? Otherwise it may not get the full list.

This processes the includes explicitly listed in the environment's manifest. The environment file has to be read before those includes are processes.

tgamblin commented on lib/spack/spack/environment/environment.py:
```diff
@@ -3061,6 +3009,119 @@ def use_config(self):
+
+        # treat relative paths as relative to the environment
+        if not os.path.isabs(config_path):
+            config_path = resolve_relative(config_path)
```
I guess I did not notice when this was added originally, but `spack.util.path,canonicalize_path` already handles relative path resolution for paths that came from config files (syaml_str already has marks on it to indicate the file provenance), so this can be simplified.  See https://github.com/spack/spack/blob/f13d998/lib/spack/spack/util/path.py#L251-L256.

tgamblin commented on lib/spack/spack/environment/environment.py:
```diff
@@ -3061,6 +3009,119 @@ def use_config(self):
+    config_stage_dir: str,
+    resolve_relative: Callable[[str], str],
+    required: List[str] = [],
+) -> List[spack.config.ConfigScope]:
```
This should all move to `spack.config`, and should add scopes under config scopes as we load them.

If the remote fetch part is too complex to move there (it may require some separate staging mechanism, which would add complexity), then I would only allow remote includes from within environments.  I think it makes sense to limit it as we work out the details.

tgamblin reviewed (changes requested):
Right now this makes `include:` into config but only sort of -- really only for environments.  So it won't solve the use case it was intended for, which is getting rid of OS-specific include directories in our config scopes.

Example:

```console
> cat ~/disable_verify.yaml
config:
  verify_ssl: false
```

Ok now I make an env that includes this:

```console
> spack env create ./test
==> Created independent environment in: /Users/gamblin2/test
==> Activate with: spack env activate ./test
> e test/spack.yaml
> cat test/spack.yaml
spack:
  # add package specs to the `specs` list
  specs: []
  view: true
  concretizer:
    unify: true
  include:
    - ~/disable_verify.yaml
> spack -e ./test config blame config | grep verify_ssl
/Users/gamblin2/disable_verify.yaml:2                                     verify_ssl: False
```

But if I do this instead:

```console
> cat ~/.spack/include.yaml
include:
  - ~/disable_verify.yaml
```

Spack doesn't actually notice it:

```console
> spack config blame config | grep verify_ssl
/Users/gamblin2/Workspace/src/spack/etc/spack/defaults/config.yaml:101    verify_ssl: True
```

So this doesn't actually make `include:` into a proper config section... it just changes how it's handled in environments.

tldahlgren committed:
Remove setting config's additionalProperties to False (per Todd)

tldahlgren committed:
Restore config::update_config_with_includes; move included_config_scopes, scopes_from_paths to config

greenc-FNAL commented:
Related: we've been trying to reuse already-built software in the following way:
1. We have a fully-concretized, built, and installed "parent" environment X.
2. We have a daughter environment Y that incorporates X via `include_concrete` of its `spack.lock` file, and then specifies different roots that may or may not coincide or be part of the dependency tree of the roots of X.
Unfortunately, concretizing Y can lead to the rebuilding of packages built for X because the `packages:` clauses (virtual package providers, global and per-package requirements, variants, etc.), and other non-spec clauses (e.g. `concretizer:`) are not honored. To resolve this we are currently splicing non-root entries from the `spack.yaml` of X into Y where they don't conflict with existing specifications in order to avoid rebuilding huge swaths of packages unnecessarily. It would be nice to be able to have the desired non-root sections of X honored in Y on a, "unless I override" basis.

tldahlgren committed:
A pass over circular dependencies

tldahlgren committed:
test_env_include_configs: tweak spack.yaml string

tldahlgren committed:
includes: update docs

tldahlgren committed:
Restore separate env and include processing to fix env tests

kwryankrattiger commented:
Noting this here as well #44285

greenc-FNAL commented:
@tldahlgren @tgamblin Where are we with this, please?

tldahlgren commented:
> @tldahlgren @tgamblin Where are we with this, please?

Was side-tracked by other work. Plan to resume this PR starting tomorrow.

tldahlgren committed:
Rebase cleanup: same diffs as before the rebase

tldahlgren committed:
Post-rebase style updates

tldahlgren committed:
Move offending imports into update_config_with_includes (but still circular)

tldahlgren committed:
Fix style issues

tldahlgren committed:
Fix pylint issues with unspecified encoding

tldahlgren committed:
Moved scopes_from_paths from environment.py to config.py

tldahlgren committed:
Move update_config_with_includes from config to main

tldahlgren committed:
environment: remove TODO comment from included_config_scopes

tldahlgren committed:
scopes_from_paths: replace resolve_relative (callable) with relative_root

tldahlgren committed:
eval_conditional: move from environment to spec; make public function

tldahlgren committed:
Refactor to create config.included_path and add OptionalPath

tldahlgren committed:
Fix style issues

tldahlgren committed:
Update fish completion

tldahlgren committed:
config.scopes_from_paths: add docstring spacing for readthedocs

tldahlgren committed:
Update the copyright lines for the new files

tldahlgren committed:
schemas: restore literal lines since new copyright is one-liner

tldahlgren committed:
_concretize_together_where_possible: Remove commented out import

tldahlgren commented:
The only thing failing at this point is patch coverage *but* the offending files are not ones changed in this PR.

tldahlgren commented:
@tgamblin Is more work needed to get the basics here supported? Or can some of your remaining issues (eg., refactoring conditional/optional handling for config versus environment files; revisit use of spack.util.path.canonicalize_path) be deferred to a follow-on PR? 

Also, should this be merged before or after #48420 ?

tldahlgren committed:
Replace scopes_from_paths with include_path_scope

tldahlgren commented:
Given `$HOME/.spack/include.yaml` for 36742f3 containing: 

```
include:
- $HOME/disable_verify.yaml
- $HOME/more.yaml
- https://raw.githubusercontent.com/spack/spack-configs/refs/heads/main/NREL/configs/mac/config.yaml
```

`more.yaml` with:

```
config:
    dirty: False
```

and `$HOME/disable_verify.yaml`:

```
config:
  verify_ssl: false
```

I get the following:

```
$ spack config blame include
---                                    include:
/g/g21/dahlgren/.spack/include.yaml:3  - $HOME/disable_verify.yaml
/g/g21/dahlgren/.spack/include.yaml:4  - $HOME/more.yaml
/g/g21/dahlgren/.spack/include.yaml:5  - https://raw.githubusercontent.com/spack/spack-configs/refs/heads/main/NREL/configs/mac/config.yaml

$ spack config blame config
---                                                                  config:
$misc_cache_location/cache/includes/config.yaml:2    build_stage:
$misc_cache_location/cache/includes/config.yaml:3    - $spack/var/spack/stage
$spack/etc/spack/defaults/config.yaml:70                     - $tempdir/$user/spack-stage
$spack/etc/spack/defaults/config.yaml:71                      - $user_cache_path/stage
$spack/cache/cache/includes/config.yaml:7                  build_jobs: 8
$HOME/more.yaml:2                                                         dirty: False
$HOME/disable_verify.yaml:2                                          verify_ssl: False
$spack/etc/spack/defaults/config.yaml:19                     install_tree:
...
```

tldahlgren committed:
main: remove unused import

tldahlgren committed:
Moved remote config retrieval out of config.py

tldahlgren committed:
Resolve style issues plus add include_path_scope return to docstring

tldahlgren committed:
test_config_change_existing: do NOT use normal misc cache path

tldahlgren committed:
Snapshot: added test_include_duplicate_source; some config cleanup

tldahlgren committed:
config: remove unused remote_path

tldahlgren committed:
test_env_with_included_config_precedence: fix docstring typo

tldahlgren committed:
Refactor retention and handling of sources associated with included paths

tldahlgren committed:
Cache web-specific config settings in util.web_config; move RemoteFileError

tldahlgren committed:
Cleanup prior; add remote_file_cache.py

tldahlgren committed:
Ignore type hint for passing config.CONFIG into web_config.update()

tldahlgren committed:
Move path processing from config to remote file utility

tldahlgren committed:
Fix readthedocs FetchError complaint

tldahlgren committed:
Restore environment include ordering lost in refactor

tldahlgren committed:
Add information on path options for global includes

tldahlgren committed:
test_updated_completion_scripts: change message to exclude global include scopes

tldahlgren committed:
Add missing colon to exception in docstring

tldahlgren commented:
Closing since superseded by #48784 .
